compiler/rustc_codegen_llvm/src/consts.rs

```diff
@@ -353,6 +361,7 @@ impl StaticMethods for CodegenCx<'ll, 'tcx> {
 
                 llvm::LLVMRustSetLinkage(new_g, linkage);
                 llvm::LLVMRustSetVisibility(new_g, visibility);
+                llvm::LLVMRustSetDSOLocal(new_g, dso_local);
```

Comment:
The "other piece of code" is here. Since we're defining the static here, I think you can, and probably should, just unconditionally set dso_local based on `should_assume_dso_local` here.

I _think_.

Reply:
Do things still work if this is moved outside of this conditional (e.g. to near `set_global_alignment` call below)?

Reply:
Testsuite passes locally at least. But I have to query linkage/visibility again or pull [these lines](https://github.com/rust-lang/rust/blob/961ba357f95cd233b7712d7784c2a75cd31a3c5a/compiler/rustc_codegen_llvm/src/consts.rs#L351-L352) before the if block making it a little bit harder to follow. Should I go forward with the change or were you just wondering what if?

Reply:
I think we should go ahead with it. I'm quite worried that there may be places in the codegen where a static is declared with the right type, but misses setting `dso_local`, in which case this branch would not be hit.

---

On another note is I think the code should look a more like:

```
if self.should_assume_dso_local(...) { 
     llvm::LLVMRustSetDSOLocal(g, true);
}
```

There's almost never a reason to unset the flag and the part of the backend that has declared the static may have a better idea as to whether the static can be a `dso_local` based on factors that `should_assume_dso_local` can't consider.

---

The fact that it is necessary to get the linkage/visibility back from LLVM I think is indicative of some larger refactor being necessary here so that the linkage/visibility derived from the Rust source is available more directly here. But that's out of scope for this PR. I thin it is perfectly fine to do whatever necessary to obtain the information you need to pass into `should_assume_dso_global` here.

Reply:
Good thinking. And indeed if I apply the following patch on top of my latest commit we have a few tests which trigger the assertion.
<details>
<summary>Patch</summary>

```diff
diff --git a/compiler/rustc_codegen_llvm/src/consts.rs b/compiler/rustc_codegen_llvm/src/consts.rs
index 245842df1b0..2748fbe0190 100644
--- a/compiler/rustc_codegen_llvm/src/consts.rs
+++ b/compiler/rustc_codegen_llvm/src/consts.rs
@@ -361,6 +361,12 @@ fn codegen_static(&self, def_id: DefId, is_mutable: bool) {
                 llvm::LLVMRustSetLinkage(new_g, linkage);
                 llvm::LLVMRustSetVisibility(new_g, visibility);
 
+                let linkage = base::linkage_from_llvm(linkage);
+                let visibility = base::visibility_from_llvm(visibility);
+                if self.should_assume_dso_local(linkage, visibility) {
+                    llvm::LLVMRustSetDSOLocal(g, true);
+                }
+
                 // To avoid breaking any invariants, we leave around the old
                 // global for the moment; we'll replace all references to it
                 // with the new global later. (See base::codegen_backend.)
@@ -372,9 +378,10 @@ fn codegen_static(&self, def_id: DefId, is_mutable: bool) {
 
             let linkage = base::linkage_from_llvm(llvm::LLVMRustGetLinkage(g));
             let visibility = base::visibility_from_llvm(llvm::LLVMRustGetVisibility(g));
-            if self.should_assume_dso_local(linkage, visibility) {
-                llvm::LLVMRustSetDSOLocal(g, true);
-            }
+            assert_eq!(
+                llvm::LLVMRustIsDSOLocal(g),
+                self.should_assume_dso_local(linkage, visibility)
+            );
 
             // As an optimization, all shared statics which do not have interior
             // mutability are placed into read-only memory.
diff --git a/compiler/rustc_codegen_llvm/src/llvm/ffi.rs b/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
index 7fb978d61dc..2b515983c6e 100644
--- a/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
+++ b/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
@@ -1014,6 +1014,7 @@ pub fn LLVMConstExtractValue(
     pub fn LLVMSetSection(Global: &Value, Section: *const c_char);
     pub fn LLVMRustGetVisibility(Global: &Value) -> Visibility;
     pub fn LLVMRustSetVisibility(Global: &Value, Viz: Visibility);
+    pub fn LLVMRustIsDSOLocal(Global: &Value) -> bool;
     pub fn LLVMRustSetDSOLocal(Global: &Value, is_dso_local: bool);
     pub fn LLVMGetAlignment(Global: &Value) -> c_uint;
     pub fn LLVMSetAlignment(Global: &Value, Bytes: c_uint);
diff --git a/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp b/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
index 2e135fbe2bd..0fd19d55903 100644
--- a/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
+++ b/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
@@ -1648,6 +1648,10 @@ extern "C" void LLVMRustSetVisibility(LLVMValueRef V,
   LLVMSetVisibility(V, fromRust(RustVisibility));
 }
 
+extern "C" bool LLVMRustIsDSOLocal(LLVMValueRef Global) {
+  return unwrap<GlobalValue>(Global)->isDSOLocal();
+}
+
 extern "C" void LLVMRustSetDSOLocal(LLVMValueRef Global, bool is_dso_local) {
   unwrap<GlobalValue>(Global)->setDSOLocal(is_dso_local);
 }
```

</details>

```
failures:
    [ui] ui/issues/issue-33992.rs
    [ui] ui/linkage-attr/linkage-detect-extern-generated-name-collision.rs
    [ui] ui/linkage-attr/linkage-detect-local-generated-name-collision.rs
    [ui] ui/type-alias-impl-trait/bound_reduction.rs#full_tait
    [ui] ui/type-alias-impl-trait/bound_reduction.rs#min_tait
```